fmriprep/interfaces/patches.py

```diff
@@ -12,49 +12,110 @@
 
 from numpy.linalg.linalg import LinAlgError
 from nipype.algorithms import confounds as nac
+from nipype.interfaces.base import (File, traits)
+from nipype.interfaces.mixins import reporting
 
 
-class RobustACompCor(nac.ACompCor):
-    """
-    Runs aCompCor several times if it suddenly fails with
-    https://github.com/poldracklab/fmriprep/issues/776
+class RetryCompCorInputSpecMixin(reporting.ReportCapableInputSpec):
+    out_report = File('report.html', usedefault=True, hash_files=False,
+                      desc='filename for warning HTML snippet')
+    # 'NaN` by default
+    failure_mode = traits.Enum(
+        'NaN', 'error',
+        usedefault=True,
+        desc='When no components are found or convergence fails, raise an error '
+             'or silently return columns of NaNs.')
 
-    """
 
+class RetryCompCorMixin(reporting.ReportCapableInterface):
     def _run_interface(self, runtime):
+        warn = self.inputs.failure_mode == 'NaN'
+
         failures = 0
+        save_exc = None
         while True:
+            success = True
+            # Identifiy success/failure in both error and NaN mode
             try:
-                runtime = super(RobustACompCor, self)._run_interface(runtime)
+                runtime = super()._run_interface(runtime)
+                if warn and self._check_nans():
+                    success = False
+            except LinAlgError as exc:
+                success = False
+                save_exc = exc
+
+            if success:
                 break
-            except LinAlgError:
-                failures += 1
-                if failures > 10:
-                    raise
-                start = (failures - 1) * 10
-                sleep(randint(start + 4, start + 10))
+
+            failures += 1
+            if failures > 10:
+                if warn:
+                    break
+                raise save_exc
+            start = (failures - 1) * 10
+            sleep(randint(start + 4, start + 10))
 
         return runtime
 
+    def _check_nans(self):
+        import numpy as np
+        outputs = self._list_outputs()
+        components = np.loadtxt(outputs['components_file'], skiprows=1)
+        return np.isnan(components).all()
+
+    def _generate_report(self):
+        snippet = '<!-- {} completed without error -->'.format(self._header)
+        if self._check_nans():
+            snippet = '''\
+<p class="elem-desc">
+    Warning: {} components could not be estimated, due to a linear algebra error.
+    While not definitive, this may be an indication of a poor mask.
```

Comment:
Should we add a strict check here that `mask.sum() > 1` and raise if not?: https://github.com/nipy/nipype/blob/fbed1f3c432e80b1e199f93d17e4d8f1da232897/nipype/algorithms/confounds.py#L1155

Reply:
We could do that, but why raise an error instead of issue a warning? The whole point of this is not to crash. People who care about CompCor can worry about why, but if they don't, the masks don't matter at all.